core/src/main/scala-2/chisel3/Contract.scala

```diff
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3
+
+import chisel3.experimental.SourceInfo
+import chisel3.internal._
+import chisel3.internal.Builder.pushCommand
+import chisel3.internal.firrtl.ir._
+import scala.language.experimental.macros
+import scala.reflect.macros.whitebox
+
+object Contract {
+
+  /** Create a `contract` block with no arguments and results.
+    *
+    * Within the contract body, `RequireProperty` and `EnsureProperty` can be
+    * used to describe the pre and post-conditions of the contract. During
+    * formal verification, contracts can be used to divide large formal checks
+    * into smaller pieces by first asserting some local properties hold, and
+    * then assuming those properties hold in the context of the larger proof.
+    *
+    * @example
+    * {{{
+    * Contract() {
+    *   RequireProperty(a >= 1)
+    *   EnsureProperty(a + a >= 2)
+    * }
+    * }}}
+    */
+  def apply()(body: => Unit)(implicit sourceInfo: SourceInfo): Unit = {
+    mapped[Unit](Seq(), _ => ())(_ => body)
+  }
+
+  /** Create a `contract` block with one or more arguments and results.
+    *
+    * Outside of verification uses, the arguments passed to the contract are
+    * simply returned as contract results. During formal verification, the
```

Comment:
what are "the contract results"? what does that mean?

Reply:
This is due to the cutpoint behaviour of a contract during formal verification. When doing formal verification, the contract doesn't just forward its argument list to its results, but instead creates symbolic values as its results that are then constrained by the contract body. (Which is why the body takes arguments as well: it needs to constrain either the hardware, or the symbolic value, depending on which part you're verifying.)